File: Palace/OPDS/TPPOPDSFeed.m
#import "NSDate+NYPLDateAdditions.h"
#import "TPPAsync.h"
#import "TPPOPDSEntry.h"
#import "TPPOPDSLink.h"
#import "TPPOPDSRelation.h"
#import "TPPSession.h"
#import "TPPXML.h"
#import "Palace-Swift.h"
#import "TPPOPDSFeed.h"

#if defined(FEATURE_DRM_CONNECTOR)
#import <ADEPT/ADEPT.h>
#endif

@interface TPPOPDSFeed ()

@property (nonatomic) NSArray *entries;
@property (nonatomic) NSString *identifier;
@property (nonatomic) NSArray *links;
@property (nonatomic) NSString *title;
@property (nonatomic) TPPOPDSFeedType type;
@property (nonatomic) BOOL typeIsCached;
@property (nonatomic) NSDate *updated;
@property (nonatomic) NSMutableDictionary *licensor;
@property (nonatomic) NSString *authorizationIdentifier;

@end

static TPPOPDSFeedType TypeImpliedByEntry(TPPOPDSEntry *const entry)
{
  BOOL entryIsGrouped = NO;

  // NOTE: A catalog entry is an acquisition feed according to section 8 of
  // OPDS Catalog 1.1 if it contains at least one acquisition link.
  BOOL entryIsCatalogEntry = entry.acquisitions.count >= 1;

  for(TPPOPDSLink *const link in entry.links) {
    if([link.rel hasPrefix:@"http://opds-spec.org/acquisition"]) {
      // This also means we have an acquisition feed.
      entryIsCatalogEntry = YES;
    } else if([link.rel isEqualToString:TPPOPDSRelationGroup]) {
      entryIsGrouped = YES;
    }
  }
  
  if(entryIsGrouped && !entryIsCatalogEntry) {
    return TPPOPDSFeedTypeInvalid;
  }
  
  return (entryIsCatalogEntry
          ? (entryIsGrouped
             ? TPPOPDSFeedTypeAcquisitionGrouped
             : TPPOPDSFeedTypeAcquisitionUngrouped)
          : TPPOPDSFeedTypeNavigation);
}

@implementation TPPOPDSFeed

+ (void)withURL:(NSURL *)URL
shouldResetCache:(BOOL)shouldResetCache
completionHandler:(void (^)(TPPOPDSFeed *feed, NSDictionary *error))handler
{
  if(!handler) {
    @throw NSInvalidArgumentException;
  }

  __block NSURLRequest *request = nil;
  NSURLRequestCachePolicy cachePolicy;
  if (shouldResetCache) {
    cachePolicy = NSURLRequestReloadIgnoringLocalCacheData;
  } else {
    cachePolicy = NSURLRequestUseProtocolCachePolicy;
  }

  if (URL == nil) {
    [TPPErrorLogger logErrorWithCode:TPPErrorCodeNoURL
                              summary:@"NYPLOPDSFeed: nil URL"
                             metadata:@{
                               @"shouldResetCache": @(shouldResetCache)
                             }];
    TPPAsyncDispatch(^{handler(nil, nil);});
    return;
  }

  request = [[[TPPNetworkExecutor shared] GET:URL
                                  cachePolicy:cachePolicy
                                  useTokenIfAvailable: NO
                                  completion:^(NSData *data, NSURLResponse *response, NSError *error) {

    if (error != nil) {
      // Note: NYPLNetworkExecutor already logged this error
      TPPAsyncDispatch(^{handler(nil, error.problemDocument.dictionaryValue);});
      return;
    }

    if (data == nil) {
      [TPPErrorLogger logErrorWithCode:TPPErrorCodeOpdsFeedNoData
                                summary:@"NYPLOPDSFeed: no data from server"
                               metadata:@{
                                 @"Request": [request loggableString],
                                 @"Response": response ?: @"N/A",
                               }];
      TPPAsyncDispatch(^{handler(nil, nil);});
      return;
    }

    if ([response isKindOfClass:[NSHTTPURLResponse class]]) {
      NSHTTPURLResponse *httpResp = (NSHTTPURLResponse*)response;

      if (httpResp.statusCode < 200 || httpResp.statusCode > 299) {
        // this captures a situation where (e.g.) borrow requests to the
        // Brooklyn lib come back with a 500 status code, no error, and non-nil
        // data containing "An internal error occurred" plain text.
        NSString *msg = [NSString stringWithFormat:@"Got %ld HTTP status with no error object.", (long)httpResp.statusCode];

        NSString *dataString = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];

        NSDictionary *problemDocDict = nil;
        if (response.isProblemDocument) {
          problemDocDict = [NSJSONSerialization JSONObjectWithData:data options:(NSJSONReadingOptions)0 error:nil];
        }

        [TPPErrorLogger logNetworkError:error
                                   code:TPPErrorCodeApiCall
                                 summary:@"NYPLOPDSFeed: HTTP response error"
                                 request:request
                                response:response
                                metadata:@{
                                  @"receivedData": dataString ?: @"N/A",
                                  @"receivedDataLength (bytes)": @(data.length),
                                  @"problemDoc": problemDocDict ?: @"N/A",
                                  @"context": msg ?: @"N/A"
                                }];

        TPPAsyncDispatch(^{handler(nil, problemDocDict);});
        return;
      }
    }
    
    TPPXML *const feedXML = [TPPXML XMLWithData:data];
    if(!feedXML) {
      TPPLOG(@"Failed to parse data as XML.");
      [TPPErrorLogger logErrorWithCode:TPPErrorCodeFeedParseFail
                                summary:@"NYPLOPDSFeed: Failed to parse data as XML"
                               metadata:@{
                                 @"request": request.loggableString,
                                 @"response": response ?: @"N/A",
                               }];
      // this error may be nil
      NSDictionary *error = [NSJSONSerialization JSONObjectWithData:data options:(NSJSONReadingOptions)0 error:nil];
      TPPAsyncDispatch(^{handler(nil, error);});
      return;
    }
    
    TPPOPDSFeed *const feed = [[TPPOPDSFeed alloc] initWithXML:feedXML];
    if(!feed) {
      TPPLOG(@"Could not interpret XML as OPDS.");
      [TPPErrorLogger logErrorWithCode:TPPErrorCodeOpdsFeedParseFail
                                summary:@"NYPLOPDSFeed: Failed to parse XML as OPDS"
                               metadata:@{
                                 @"request": request.loggableString,
                                 @"response": response ?: @"N/A",
                               }];
      TPPAsyncDispatch(^{handler(nil, nil);});
      return;
    }
    
    TPPAsyncDispatch(^{handler(feed, nil);});
  }] originalRequest];
}

- (instancetype)initWithXML:(TPPXML *const)feedXML
{
  self = [super init];
  if(!self) return nil;
  
  if(!feedXML) {
    return nil;
  }
  
  // Sometimes we get back JUST an entry, and in that case we just want to construct a feed with
  // nothing set other than the entry.
  if ([feedXML.name isEqual:@"entry"]) {
    TPPOPDSEntry const* entry = [[TPPOPDSEntry alloc] initWithXML:feedXML];
    if (entry) {
      self.entries = @[entry];
      return self;
    } else {
      TPPLOG(@"Error creating single OPDS entry from feed.");
      return nil;
    }
  }
  
  if(!((self.identifier = [feedXML firstChildWithName:@"id"].value))) {
    TPPLOG(@"Missing required 'id' element.");
    return nil;
  }
  
  {
    NSMutableArray *const links = [NSMutableArray array];
    
    for(TPPXML *const linkXML in [feedXML childrenWithName:@"link"]) {
      TPPOPDSLink *const link = [[TPPOPDSLink alloc] initWithXML:linkXML];
      if(!link) {
        TPPLOG(@"Ignoring malformed 'link' element.");
        continue;
      }
      [links addObject:link];
    }
    
    self.links = links;
  }
  
  if(!((self.title = [feedXML firstChildWithName:@"title"].value))) {
    TPPLOG(@"Missing required 'title' element.");
    return nil;
  }
  
  {
    NSString *const updatedString = [feedXML firstChildWithName:@"updated"].value;
    if(!updatedString) {
      TPPLOG(@"Missing required 'updated' element.");
      return nil;
    }
    
    self.updated = [NSDate dateWithRFC3339String:updatedString];
    if(!self.updated) {
      TPPLOG(@"Element 'updated' does not contain an RFC 3339 date.");
      return nil;
    }
  }
  
  {
    NSMutableArray *const entries = [NSMutableArray array];
    
    for(TPPXML *const entryXML in [feedXML childrenWithName:@"entry"]) {
      TPPOPDSEntry *const entry = [[TPPOPDSEntry alloc] initWithXML:entryXML];
      if(!entry) {
        TPPLOG(@"Ingoring malformed 'entry' element.");
        continue;
      }
      [entries addObject:entry];
    }
    
    self.entries = entries;
  }
  
  {
    TPPXML *patronXML = [feedXML firstChildWithName:@"patron"];
    if (patronXML && patronXML.attributes.allValues.count>0) {
      NSString *barcode = patronXML.attributes[@"simplified:authorizationIdentifier"];
      self.authorizationIdentifier = barcode;
    }
  }
  
  {
    TPPXML *licensorXML = [feedXML firstChildWithName:@"licensor"];
    if (licensorXML && licensorXML.attributes.allValues.count>0) {
      NSString *vendor = licensorXML.attributes[@"drm:vendor"];
      TPPXML *tokenXML = [licensorXML firstChildWithName:@"clientToken"];
      
      if (tokenXML) {
        NSString *clientToken = tokenXML.value;
        self.licensor = @{@"vendor":vendor,
                          @"clientToken":clientToken}.mutableCopy;
      } else {
        TPPLOG(@"Licensor not saved. Error parsing clientToken into XML.");
      }
    } else {
      TPPLOG(@"No Licensor found in OPDS feed. Moving on.");
    }
  }
  
  return self;
}

- (TPPOPDSFeedType)type
{
  if(self.typeIsCached) {
    return _type;
  }
  
  self.typeIsCached = YES;
  
  if(self.entries.count == 0) {
    _type = TPPOPDSFeedTypeAcquisitionUngrouped;
    return _type;
  }
  
  TPPOPDSFeedType const provisionalType = TypeImpliedByEntry(self.entries.firstObject);
  
  if(provisionalType == TPPOPDSFeedTypeInvalid) {
    _type = TPPOPDSFeedTypeInvalid;
    return _type;
  }
  
  for(unsigned int i = 1; i < self.entries.count; ++i) {
    if(TypeImpliedByEntry(self.entries[i]) != provisionalType) {
      _type = TPPOPDSFeedTypeInvalid;
      return _type;
    }
  }
  
  _type = provisionalType;
  return _type;
}

@end

File: Palace/OPDS/TPPOPDSFeed.swift
import Foundation

@objc enum TPPOPDSFeedType: Int {
  case invalid
  case acquisitionGrouped
  case acquisitionUngrouped
  case navigation
}

@objc class TPPOPDSFeed: NSObject {

  @objc private(set) var entries: [Any] = []
  @objc private(set) var identifier: String?
  @objc private(set) var links: [Any] = []
  @objc private(set) var title: String?
  @objc private(set) var updated: Date?
  @objc private(set) var licensor: NSDictionary?
  @objc private(set) var authorizationIdentifier: String?

  private var _type: TPPOPDSFeedType = .invalid
  private var typeIsCached = false

  @objc var type: TPPOPDSFeedType {
    if typeIsCached { return _type }
    typeIsCached = true

    guard let entries = self.entries as? [TPPOPDSEntry], !entries.isEmpty else {
      _type = .acquisitionUngrouped
      return _type
    }

    let provisionalType = Self.typeImplied(by: entries[0])
    if provisionalType == .invalid {
      _type = .invalid
      return _type
    }

    for i in 1..<entries.count {
      if Self.typeImplied(by: entries[i]) != provisionalType {
        _type = .invalid
        return _type
      }
    }

    _type = provisionalType
    return _type
  }

  @objc static func withURL(
    _ url: URL?,
    shouldResetCache: Bool,
    useTokenIfAvailable: Bool,
    completionHandler handler: @escaping (TPPOPDSFeed?, NSDictionary?) -> Void
  ) {
    guard let url = url else {
      TPPErrorLogger.logError(
        withCode: .noURL,
        summary: "NYPLOPDSFeed: nil URL",
        metadata: ["shouldResetCache": shouldResetCache]
      )
      TPPAsyncDispatch { handler(nil, nil) }
      return
    }

    let cachePolicy: NSURLRequest.CachePolicy = shouldResetCache ? .reloadIgnoringLocalCacheData : .useProtocolCachePolicy

    var request: URLRequest?

    let task = TPPNetworkExecutor.shared.get(
      url,
      cachePolicy: cachePolicy,
      useTokenIfAvailable: useTokenIfAvailable
    ) { data, response, error in

      if let error = error {
        TPPAsyncDispatch { handler(nil, (error as NSError).problemDocument?.dictionaryValue as NSDictionary?) }
        return
      }

      guard let data = data else {
        TPPErrorLogger.logError(
          withCode: .opdsFeedNoData,
          summary: "NYPLOPDSFeed: no data from server",
          metadata: [
            "Request": (request as NSURLRequest?)?.loggableString ?? "N/A",
            "Response": response ?? "N/A"
          ]
        )
        TPPAsyncDispatch { handler(nil, nil) }
        return
      }

      if let httpResp = response as? HTTPURLResponse,
         httpResp.statusCode < 200 || httpResp.statusCode > 299 {
        let dataString = String(data: data, encoding: .utf8)

        var problemDocDict: NSDictionary?
        if response?.isProblemDocument == true {
          problemDocDict = try? JSONSerialization.jsonObject(with: data, options: []) as? NSDictionary
        }

        TPPErrorLogger.logNetworkError(
          error,
          code: .apiCall,
          summary: "NYPLOPDSFeed: HTTP response error",
          request: request as NSURLRequest?,
          response: response,
          metadata: [
            "receivedData": dataString ?? "N/A",
            "receivedDataLength (bytes)": data.count,
            "problemDoc": problemDocDict ?? "N/A",
            "context": "Got \(httpResp.statusCode) HTTP status with no error object."
          ]
        )

        TPPAsyncDispatch { handler(nil, problemDocDict) }
        return
      }

      guard let feedXML = TPPXML.xml(withData: data) else {
        Log.log("Failed to parse data as XML.")
        TPPErrorLogger.logError(
          withCode: .feedParseFail,
          summary: "NYPLOPDSFeed: Failed to parse data as XML",
          metadata: [
            "request": (request as NSURLRequest?)?.loggableString ?? "N/A",
            "response": response ?? "N/A"
          ]
        )
        let errorDict = try? JSONSerialization.jsonObject(with: data, options: []) as? NSDictionary
        TPPAsyncDispatch { handler(nil, errorDict) }
        return
      }

      guard let feed = TPPOPDSFeed(xml: feedXML) else {
        Log.log("Could not interpret XML as OPDS.")
        TPPErrorLogger.logError(
          withCode: .opdsFeedParseFail,
          summary: "NYPLOPDSFeed: Failed to parse XML as OPDS",
          metadata: [
            "request": (request as NSURLRequest?)?.loggableString ?? "N/A",
            "response": response ?? "N/A"
          ]
        )
        TPPAsyncDispatch { handler(nil, nil) }
        return
      }

      TPPAsyncDispatch { handler(feed, nil) }
    }

    request = task?.originalRequest
  }

  @objc init?(xml feedXML: TPPXML?) {
    super.init()

    guard let feedXML = feedXML else { return nil }

    // Sometimes we get back JUST an entry
    if feedXML.name == "entry" {
      guard let entry = TPPOPDSEntry(xml: feedXML) else {
        Log.log("Error creating single OPDS entry from feed.")
        return nil
      }
      entries = [entry]
      return
    }

    guard let id = feedXML.firstChild(withName: "id")?.value else {
      Log.log("Missing required 'id' element.")
      return nil
    }
    identifier = id

    var parsedLinks = [TPPOPDSLink]()
    for linkXML in feedXML.childrenWithName("link") {
      guard let link = TPPOPDSLink(xml: linkXML) else {
        Log.log("Ignoring malformed 'link' element.")
        continue
      }
      parsedLinks.append(link)
    }
    links = parsedLinks

    guard let t = feedXML.firstChild(withName: "title")?.value else {
      Log.log("Missing required 'title' element.")
      return nil
    }
    title = t

    guard let updatedString = feedXML.firstChild(withName: "updated")?.value else {
      Log.log("Missing required 'updated' element.")
      return nil
    }
    guard let updatedDate = NSDate.date(withRFC3339String: updatedString) as Date? else {
      Log.log("Element 'updated' does not contain an RFC 3339 date.")
      return nil
    }
    updated = updatedDate

    var parsedEntries = [TPPOPDSEntry]()
    for entryXML in feedXML.childrenWithName("entry") {
      guard let entry = TPPOPDSEntry(xml: entryXML) else {
        Log.log("Ignoring malformed 'entry' element.")
        continue
      }
      parsedEntries.append(entry)
    }
    entries = parsedEntries

    if let patronXML = feedXML.firstChild(withName: "patron"),
       let attrs = patronXML.attributes as? [String: String],
       !attrs.isEmpty {
      authorizationIdentifier = attrs["simplified:authorizationIdentifier"]
    }

    if let licensorXML = feedXML.firstChild(withName: "licensor"),
       let attrs = licensorXML.attributes as? [String: String],
       !attrs.isEmpty {
      if let vendor = attrs["drm:vendor"],
         let tokenXML = licensorXML.firstChild(withName: "clientToken") {
        let clientToken = tokenXML.value
        licensor = ["vendor": vendor, "clientToken": clientToken] as NSDictionary
      } else {
        Log.log("Licensor not saved. Error parsing clientToken into XML.")
      }
    } else {
      Log.log("No Licensor found in OPDS feed. Moving on.")
    }
  }

  private static func typeImplied(by entry: TPPOPDSEntry) -> TPPOPDSFeedType {
    var entryIsGrouped = false
    var entryIsCatalogEntry = !entry.acquisitions.isEmpty

    for link in entry.links {
      if let rel = link.rel {
        if rel.hasPrefix("http://opds-spec.org/acquisition") {
          entryIsCatalogEntry = true
        } else if rel == TPPOPDSRelationGroup {
          entryIsGrouped = true
        }
      }
    }

    if entryIsGrouped && !entryIsCatalogEntry {
      return .invalid
    }

    return entryIsCatalogEntry
      ? (entryIsGrouped ? .acquisitionGrouped : .acquisitionUngrouped)
      : .navigation
  }
}
